src/V3Gate.cpp

```diff
@@ -261,535 +609,294 @@ class GateOkVisitor final : public VNVisitorConst {
+
+        const std::function<void(AstNodeVarRef*)> visit = [&](AstNodeVarRef* nodep) -> void {
+            // See if this variable has a substitution
+            AstVarScope* const vscp = nodep->varScopep();
+            const auto& it = substitutions.find(vscp);
+            if (it == substitutions.end()) return;
+
+            // Do not substitute circular logic
+            if (!replaced.insert(vscp).second) return;
+
+            // Substitute nodep with substp
+            AstNodeExpr* const substp = it->second;
+
+            UASSERT_OBJ(nodep->access().isReadOnly(), nodep, "Can't replace write references");
+            UASSERT_OBJ(!VN_IS(substp, NodeVarRef) || !nodep->isSame(substp), substp,
+                        "Replacing node with itself; perhaps circular logic?");
+            // The replacement
+            AstNodeExpr* const newp = substp->cloneTreePure(false);
+            // Which fileline() to use? If replacing with logic, an error/warning is likely to want
+            // to point to the logic IE what we're replacing with. However, a VARREF should point
+            // to the original as it's otherwise confusing to throw warnings that point to a PIN
+            // rather than where the pin us used.
+            if (VN_IS(newp, VarRef)) newp->fileline(nodep->fileline());
+            // Make the newp an rvalue like nodep.
+            if (AstNodeVarRef* const varrefp = VN_CAST(newp, NodeVarRef)) {
+                varrefp->access(VAccess::READ);
+            }
+            // Replace the node
+            nodep->replaceWith(newp);
+            VL_DO_DANGLING(nodep->deleteTree(), nodep);
+            // Recursively substitute the new tree
+            newp->foreach(visit);
+
+            // Remove from recursion filter
+            replaced.erase(vscp);
+        };
+
+        logicp->foreach(visit);
+
+        AstNode* const simplifiedp = V3Const::constifyEdit(logicp);
+        UASSERT_OBJ(simplifiedp == logicp, simplifiedp, "Should not remove whole logic");
+        for (const auto& pair : substitutions) pair.second->deleteTree();
+        substitutions.clear();
     }
 
-public:
-    // CONSTRUCTORS
-    explicit GateVisitor(AstNode* nodep) { iterate(nodep); }
-    ~GateVisitor() override {
-        V3Stats::addStat("Optimizations, Gate sigs deleted", m_statSigs);
-        V3Stats::addStat("Optimizations, Gate inputs replaced", m_statRefs);
-        V3Stats::addStat("Optimizations, Gate sigs deduped", m_statDedupLogic);
-        V3Stats::addStat("Optimizations, Gate assign merged", m_statAssignMerged);
+    static GateVarVertex* ffToVarVtx(V3GraphVertex* vtxp) {
+        while (vtxp && !vtxp->is<GateVarVertex>()) vtxp = vtxp->verticesNextp();
+        return static_cast<GateVarVertex*>(vtxp);
     }
-};
 
-//----------------------------------------------------------------------
+    template <bool AllowMultiIn>
+    void optimizeSignals() {
```

Comment:
It's just my visceral dislike for loop invariant conditionals inside loops, which I often find in profiler reports. On this one though you are right, there is enough stuff in that loop that this would likely make no difference so reverted.

One nice property of using the template argument is that they show up as separate functions in profiles, so you can judge the contribution of each variant. Not important in this case though.